File: Baekjoon/donggeun/소형기관차.py
import sys
input = sys.stdin.readline

n = int(input())
nums = list(map(int,input().split()))
m = int(input())

pre_fix = [0]
dp = [[0]*(n+1) for _ in range(4)]
for i in nums:
    pre_fix.append(pre_fix[-1] + i)

for i in range(1, 4):
    for j in range(i*m, n+1):
        dp[i][j] = max(dp[i][j-1], dp[i-1][j-m] + pre_fix[j]-pre_fix[j-m])

print(dp[3][n])

File: Baekjoon/joowan/소형기관차.js
const { Console } = require('console');
const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : __dirname + '/input.txt';
let input = fs.readFileSync(filePath).toString().split('\n');

function solution() {
  let N = +input[0];
  let people = input[1].split(' ').map(Number);
  let M = +input[2];
  let sumPeople = [0];
  let dp = Array.from({ length: 4 }, () => Array(N + 1).fill(0));
  for (let i = 0; i < N; i++) {
    sumPeople.push(sumPeople[i] + people[i]);
  }

  for (let i = 1; i <= 3; i++) {
    for (let j = i * M; j <= N; j++) {
      dp[i][j] = Math.max(dp[i][j - 1], dp[i - 1][j - M] + (sumPeople[j] - sumPeople[j - M]));
    }
  }
  console.log(dp[3][N]);
}
solution();
